### PR TITLE
Update email addresses on all types of hearings

### PR DESCRIPTION
**Pull request by odonnadieu**
Resolves [CASEFLOW-2102](https://vajira.max.gov/browse/CASEFLOW-2102)

**PR Stack**

Part 1
[Part 2](https://github.com/department-of-veterans-affairs/caseflow/pull/16894)

### Description

Updates the hearing details page to include the email fields for all hearing types. These fields are optional for types other than "Virtual", but can always be changed. 

Part of the work for this feature included refactoring the way that emails are updated and retrieved from the backend. There were additional layers added to handle the complex relationship between hearings, virtual hearings, and email recipients.

### Acceptance Criteria
- [ ] Code compiles correctly

### Testing Plan

The following workflows have been effected by this change:

- Edit Hearing Details (Daily Docket)
- Edit Hearing Details (Details)
- Convert Hearing Type

### User Facing Changes
 - [ ] Screenshots of UI changes added to PR & Original Issue

 BEFORE|AFTER
 ---|---

### CODE REVIEW

File: app/controllers/hearings_controller.rb [hunk 1]
```diff
@@ -110,6 +110,10 @@ def hearing_external_id
     :appellant_email, :judge_email, :representative_email, :request_cancelled, :appellant_tz, :representative_tz
   ].freeze
 
+  EMAIL_RECIPIENTS_ATTRIBUTES = [
```
Comment (nweinmeister) [inline app/controllers/hearings_controller.rb:113]: Addressing CC issue `HearingsController has 6 constants`, would it make more sense to put these attr constants in the models that they contain the attributes for? Then reference them like `HearingEmailRecipient::ATTRIBUTES` or something like that?

File: app/models/hearings/forms/base_hearing_update_form.rb [hunk 2]
```diff
@@ -228,6 +250,24 @@ def define_virtual_hearing_updates
     updates
   end
 
+  def email_sent_updates!
```
Comment (charleyf) [inline app/models/hearings/forms/base_hearing_update_form.rb:253]: The `appellant_email_sent_flag` method already checks `appellant_recipient.present?` so these can be much simpler I think.

File: app/models/concerns/has_hearing_email_recipients_concern.rb [hunk 3]
```diff
@@ -4,15 +4,14 @@ module HasHearingEmailRecipientsConcern
   extend ActiveSupport::Concern
```
Comment (charleyf) [inline app/models/concerns/has_hearing_email_recipients_concern.rb:4]: I have some ideas about how to make these getter methods like `appellant_email_address` clearer. Happy to walk through these changes as well. Here's a PR (which will never be merged) with the changes I'm suggesting.

https://github.com/department-of-veterans-affairs/caseflow/pull/16911/files

**Review by charleyf — COMMENTED** [review]
I've reviewed the backend of this and have some potential revisions in the comments of this review. I'm moving on to reviewing the frontend now.

**Review by charleyf — COMMENTED** [review]
Frontend review is now up as well.

File: app/models/hearings/forms/base_hearing_update_form.rb [hunk 4]
```diff
@@ -19,31 +19,25 @@ class BaseHearingUpdateForm
...
-      end
+      add_update_hearing_alert if show_update_alert?
+      create_or_update_virtual_hearing if should_create_or_update_virtual_hearing?
 
       after_update_hearing
     end
+    # reload hearing so new virtual hearing changes are visible
+    hearing.reload
 
-    if virtual_hearing_changed
```
[the remark this answers is not in the document]
Comment (charleyf) [reply app/models/hearings/forms/base_hearing_update_form.rb:39]: @sahalliburton, Everything else on this looks ignorable, but do you know if we have a test that would catch if we accidentally more of these start_async_jobs ?

[the remark this answers is not in the document]
Comment (sahalliburton) [reply app/models/hearings/forms/base_hearing_update_form.rb:39]: @charleyf yes I believe so, part of the test checks whether it starts the async job correctly

[the remark this answers is not in the document]
Comment (charleyf) [reply app/models/hearings/forms/base_hearing_update_form.rb:39]: Makes sense. I ended up adding a test that directly checks this. It's here if incorporating it makes sense ( I suspect it doesn't given the time constraints ).

**Review by charleyf — APPROVED** [review]
I've gone back through my earlier comments and with the work Abby's been doing on tests I think that this PR is ready.

**Review by nweinmeister — APPROVED** [review]

**Pull request merged**